### PR TITLE
Feat: Add creator fields as filter options

### PR DESCRIPTION
**Pull request by Feranchz**
### What does it do?

1. Two new filters, "Created by" and "Updated by," have been added to the list view in the content manager.

### How to test it?

- You should be able to select a "Created by" or "Updated by" filter and filter by them. The tag should contain the user display name
- You should be able to enter directly to a url with filters (example: /admin/content-manager/collectionType/api::address.address?page=1&pageSize=10&sort=postal_code:ASC&filters[$and][0][createdBy][id][$eq]=1) and see the tags with the display name of every user correctly

### CODE REVIEW

**Comment by Marc-Roig** [comment]
Good job :) This looks like a more cleaner implementation than before.

I have a couple of comments before I look at the code implementation:
- I can not save entities for other users than super admin, when I press save I see this error
image

- The new createdBy & updatedBy permission is not added by default to the other users that are not superadmins
image
So new users will no longer see those fields and the super admins will have to manually update the permissions. @innerdvations do you know if this can be done automatically?

- It seems to me we have too many ways to make fields visible, invisible, hidden, filterable, ...
	- For me visible should control the visibility of the attribute in the UI, not if it's filterable or not.
	- It would be more explicit to me to:
		- Have another "filterable" attribute in the schema
		- Make all visible fields filterable and sortable, but add the ability to set its permissions the same as you did here for created by and updated by.


I wouldn't mind continue with this implementation as long as we continue pushing this forward to find a good generic solution.

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js:12]: I'm not sure about `staleTime` here. The value seems rather arbitrary. Could you explain your thinking behind it?

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js:13]: I'd prefer if you could inline this piece in the jsx.

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: @Marc-Roig I'd be interested in your opinion here, because I lack some knowledge: I think at the moment the creator fields are excluded from RBAC, so that e.g. users who can't see all fields of a content-type, can always see who has created/ edited an entity.

From what I understand now is that we are changing this, so that these fields are covered by RBAC. That also means we can not assume a user has always access to them?

If so then this seems to be a big deal to me and I think I'd like to involve others to discuss the implications of that ...

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js:13]: Could you please rename this to `CREATOR_ATTRIBUTES` to be consistent?

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js:43]: Do you think we need both checks? `target === 'admin::user'` seems sufficient for me.

File: packages/core/content-manager/server/services/utils/configuration/attributes.js
Comment (gu-stav) [inline packages/core/content-manager/server/services/utils/configuration/attributes.js:143]: Should these defined as a constant somewhere?

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js
Comment (Marc-Roig) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: Adding @innerdvations into the discussion. I do think creator attributes should be seen by everybody by default too. If I recall correctly the problem appears when you make those fields filterable.

Nevertheless, apart from people not seeing the fields, which other problems do you think could appear here? I can only think of Audit logs not showing the creator fields now.

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js
Comment (Marc-Roig) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js:43]: does this affect to all relations to admin::users? there might be more relations pointing to that Content Type (Review workflow Assignee, other people custom relations...)

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: > Nevertheless, apart from people not seeing the fields, which other problems do you think could appear here? I can only think of Audit logs not showing the creator fields now.

I don't know tbh, but from discussions with Alex I had the feeling these fields were deliberately excluded from RBAC, because even if a user doesn't have access to other users, it might still be important to understand who has created/ edited something else.

I think I would prefer:

- @Feranchz if you could document in the PR description the discussion you had about that to understand the thinking behind it
- if this change could be documented using an ADR, which can then be discussed in the sync meetings

This feels like a small change, but I have the feeling there are widespread implications and I'd like to make sure the whole team is on board with that.

Comment (innerdvations) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: I agree we need an ADR or RFC for this, which should very clearly define:

- if `visible` just means hidden from display, those fields must still be shown in RBAC or there needs to be an explicit private/public [the direct problem this PR is solving]
- what `private` and `visible` mean
- should we add a `readable` or `public` field that means it's always available regardless of permissions?
- which creator fields should be publicly available (or not)?
- based on all of that, what should appear in the RBAC permissions?

Comment (Boegie19) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: creator fields should fall under rbac.
Auditlogs could just bypass rbac to get the right information (Since it is a system tool and not user)

My reasoning is it adds an extra security layer aka you can then just say with rbac the following fields this user should never be allowed to populate example would be resetToken then instad of having only the the 1 layer of it being senatized you have 2 secutiy layers. what means if one fails the other will still be there. like if rbac is setup properly aka it checks populates filters and sorts then RBAC is a good second layer to have over everything.

Comment (Marc-Roig) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: We might remove the extra security layer if we do bypasses imo, and make it confusing for us. I agree that having the creator fields in RBAC + all other non visible fields could be a great idea, and we need to see the implications of it.

Comment (Boegie19) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: @Marc-Roig Not how I see it
auditlogs is done by the system and not the user.
RBAC is for user permission
so system tasks bypass RBAC

Comment (Marc-Roig) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: RBAC is one part that could be affected for this, but we might have more. And adding bypasses each time we need to show creator fields might lead to unexpected data leaks.

Comment (innerdvations) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: Agree they should fall under RBAC, there are many cases in which creator fields should be completely private and not available in the api response. We need to determine if we need to change anything about how it's currently working that could be a breaking change and have it all written down clearly (because right now it isn't)

Comment (Boegie19) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:52]: Again you don't do that.
You make a diffrence between user and system actions.

all system actions bypass RBAC fully since you have full cotrol over everying example would be auditlogs.
You would still use RBAC for the edit views since that is a user action and not system.

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js
Comment (gu-stav) [reply packages/core/admin/admin/src/content-manager/components/AttributeFilter/AdminUsersFilter.js:12]: Haha, ok. Let's remove it here. react-query takes care of the caching for us during rendering, and once the list of users needs to be fetched, it should be fresh (e.g. because a new user has been added since the last load).

The CM fetches more than it should in general, but therefore always delivers non-stale data, which IMO is preferred for the list-views.

File: packages/core/admin/server/services/permission/permissions-manager/sanitize.js
Comment (Marc-Roig) [inline packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: minor thing, and probably something we can work in another PR.

If now we can filter & sort by non visible fields, we should be able to set those fields as the default sort 
image

But that probably is another discussion.

**Comment by Marc-Roig** [comment]
if the api allows filtering by non visible & writtable fields, shouldn't all those fields appear in the content manager filter? 
image

File: packages/core/admin/server/services/permission/permissions-manager/sanitize.js
Comment (gu-stav) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: I agree and think we should do this as part of this PR :)

Comment (Marc-Roig) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: Agree, that is why I think specifically this could be a separate discussion. On the List view, users can only sort by fields that are visible.

What I think we should allow in the list view is filtering by any nonvisible & writable field. And if we decide to do that, the work done there might help with the "Default Sort" too.

Comment (Boegie19) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: I think that this is somting to look into when you guys do V5 doing it any sooner would make 0 sense
since we don't yet know how V5 will change this

Comment (Marc-Roig) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: We have plans to refactor this for V5, but we are pushing it because we also need it in V4 :) For features like Fernando's one and also for Review Workflows.

Comment (Boegie19) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: I see I would atleast wait until the V5 RFC is stable then we can write it in such a way that we can keep it. and that it would also for for the V5 use ceases or would work wtih v5 with minimal change.

writing it now to rewrite it in half a year makes no sence

Comment (Marc-Roig) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: I am just saying we could explore this now, that will also help us write the RFC for v5.

Comment (Boegie19) [reply packages/core/admin/server/services/permission/permissions-manager/sanitize.js:277]: I agree with that it needs to be explored but I think we should explore it with the v5 structure and then back ported to v4 if possible.

**Comment by Marc-Roig** [comment]
I have been checking pottential issues we could introduce with this change, I believe there is none.

The first concern we had was being able to filter by localizations that a role doesnt have permissions to READ. After a bit of investigation I saw we always return the localizations in the Admin API even if you don't have permissions to Read them, just that they are not displayed.

So filtering by data that is available in the API doesnt seem a concern to me.

The same logic can be applied to relations or fields that are not visible , filtering & sorting by data that you can already see in the API as a user seems OK to me.

ping @alexandrebodin @Convly

**Comment by Marc-Roig** [comment]
@Feranchz agree, that sounds like a good tradeof for now.

**Review by Convly — COMMENTED** [review]
The compromise looks ok to me now, but I wasn't able to make the feature work for some reason (the filter on createdBy was not working).

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js
Comment (joshuaellis) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/index.js:76]: this is very different to other attributes, im sure theres a reason for it – assuming there is can you write a code comment explaining why you're doing this here as opposed to including it as part of `useAllowedAttributes`? :)

**Review by joshuaellis — DISMISSED** [review]
I'm happy but lets wait for some other approvals before merging 👍🏼

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (alexandrebodin) [inline packages/core/helper-plugin/src/utils/getDisplayName/index.js:2]:
```suggestion
 * Retrieves the display name of an admin panel user
```

File: packages/core/content-manager/server/services/utils/configuration/metadatas.js
Comment (alexandrebodin) [inline packages/core/content-manager/server/services/utils/configuration/metadatas.js:108]: Curious to know what usecase we are addressing here 🤔 ?

File: packages/core/admin/admin/src/translations/en.json
Comment (alexandrebodin) [inline packages/core/admin/admin/src/translations/en.json:654]:
```suggestion
  "content-manager.components.Filters.usersSelect.label": "Search and select a user to filter by",
```

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
Everything looks to be working as expected 👍  

One thing I noticed is in the list view we display the firstname when we use the displayName in the filters. Could we use the displayName in the list view too to make all of this consistenet ?

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:36]:
```suggestion
  const attributesWithReadPermissions = readPermissionsForSlug?.[0]?.properties?.fields ?? [];
```

Not sure if all parts of the path need to be optional but it would be great if we could continue to fade out lodash where applicable.

File: packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js
Comment (gu-stav) [inline packages/core/admin/admin/src/content-manager/components/AttributeFilter/hooks/useAllowedAttributes.js:38]: Same here

File: packages/core/helper-plugin/src/index.js
Comment (gu-stav) [inline packages/core/helper-plugin/src/index.js:105]: What do you think about not sharing this utils from the helper-plugin for now, since you only need it in the CM?

This would give us some time to figure out how we want this to work so that we can apply it everywhere.

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (gu-stav) [inline packages/core/helper-plugin/src/utils/getDisplayName/index.js:16]: I think firstname is a mandatory field for users, so we don't have to check for it. I don't know if SSO can be an exception? cc @Convly ?

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (gu-stav) [inline packages/core/helper-plugin/src/utils/getDisplayName/index.js:26]: Why trim?

File: packages/core/admin/admin/src/translations/en.json
Comment (gu-stav) [reply packages/core/admin/admin/src/translations/en.json:654]: I think this is fine :)

Comment (alexandrebodin) [reply packages/core/admin/admin/src/translations/en.json:654]: I did a quick search that's not very clear

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (Convly) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:16]: With SSO you can have either firstname + lastname or username (if I recall correctly)

Comment (Convly) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:16]: On second thought, this is not the first time I get asked this, maybe we could add a comment in this utils to add this context?
What do you think?

File: packages/core/content-manager/server/services/utils/configuration/metadatas.js
Comment (Marc-Roig) [reply packages/core/content-manager/server/services/utils/configuration/metadatas.js:108]: Investigated a bit, I don't see what purpose does this serve 🤔 . I will dig a bit more and if I don't find anything I will remove it in this PR and discuss it with Fernando later.

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (Marc-Roig) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:26]: Apparently the message gets formatted as "First name "if the last name is empty.
I say the trim should not be here, as this function does not know how the formatMessage works. wdyt?

File: packages/core/helper-plugin/src/index.js
Comment (markkaylor) [reply packages/core/helper-plugin/src/index.js:105]: It could be valuable to plugin developers but I tend to agree with Gustav and leave it in the CM for now if that's currently the only place it is used. Fernando is on vacation this week but he might have an opinion as well.

**Comment by Marc-Roig** [comment]
From alex feedback:
> One thing I noticed is in the list view we display the firstname when we use the displayName in the filters. Could we use the displayName in the list view too to make all of this consistenet ?

@markkaylor @jhoward1994  could you give me a hand here? :) 

- @markkaylor  Is it intended to show the creator fields in the list view? I first understoodd the original goal of this PR was to add it in the filters.
- The displayed value in the listview is the one defined inside the configuration metadatas. And for the creator fields it's the first name. Is there any other way where we could use the displayName instead ?

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (gu-stav) [inline packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: Haven't tested it, but I think we can use the [`select`](https://formatjs.io/docs/core-concepts/icu-syntax#select-format) syntax to render the whitespace conditionally and avoid the `.trim()`

```suggestion
        defaultMessage: '{firstname}{lastname, select, undefined {} other { {lastname}}}',
```

Comment (Marc-Roig) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: apparently lastname is an empty string when it does not exist, I would say it could be a bit more work to check for possible lastname null or empty values. What do you think if we leave the trim?

**Comment by Marc-Roig** [comment]
I believe there are some decisions that will take a while for us to sort out :) . I will work in another PR to get the part that also unblocks Content Platform work. 

I will continue working on suggestions here until we merge this work / fernando comes back.

cc @alexandrebodin @gu-stav

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (gu-stav) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: Yeah, if that can happen I am in favor of that. Since you know me you probably know I'd love to find out why it is empty string over null, but ok :D

File: packages/core/helper-plugin/src/index.js
Comment (gu-stav) [reply packages/core/helper-plugin/src/index.js:105]: If this is the last open discussion I'd prefer to make the util local to the CM and move on. We can always expose more, but we can never remove something publicly exposed.

Comment (Marc-Roig) [reply packages/core/helper-plugin/src/index.js:105]: will do 👍 sorry I have been busy with other topics today, I will move the util into CM

File: packages/core/helper-plugin/src/utils/getDisplayName/index.js
Comment (alexandrebodin) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: FYI On 1st admin registration the client sends empty string instead of no value or null for the lastname when you leave the input empty and the BE allows it

Comment (gu-stav) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: Wonderful, thanks. I'll create a bugreport.

Comment (gu-stav) [reply packages/core/helper-plugin/src/utils/getDisplayName/index.js:21]: https://github.com/strapi/strapi/issues/17287

**Review by gu-stav — APPROVED** [review]
I have not done functional QA, but the code looks good to me 🚀

**Review by simotae14 — APPROVED** [review]
QA is ok, well done

**Pull request merged**
